posts/2020-11-19-Rust-1.48.md

```diff
@@ -0,0 +1,224 @@
+
+### Easier linking in rustdoc
+
+Rustdoc, the library documentation tool included in the Rust distribution,
+lets you write documentation in Markdown. This makes it very easy to use, but
+also has some pain points. Let's say that you are writing some documentation
+for some Rust code that looks like this:
+
+```rust
+pub mod foo {
+    pub struct Foo;
+}
+
+pub mod bar {
+    pub struct Bar;
+}
+```
+
+We have two modules, each with a struct inside. Imagine we wanted to use these
+two structs together; we may want to note this in the documentation. So we'd
+write some docs that look like this:
+
+```rust
+pub mod foo {
+    /// Some docs for `Foo`
+    ///
+    /// You may want to use `Foo` with `Bar`.
+    pub struct Foo;
+}
+
+pub mod bar {
+    /// Some docs for `Bar`
+    ///
+    /// You may want to use `Bar` with `Foo`.
+    pub struct Bar;
+}
+```
+
+That's all well and good, but it would be really nice if we could link to these
+other types. That would make it much easier for the users of our library to
+navigate between them in our docs.
+
+The problem here is that Markdown doesn't know anything about Rust, and the
+URLs that rustdoc generates. So what Rust programmers have had to do is write
+those links out manually:
+
+```rust
+pub mod foo {
+    /// Some docs for `Foo`
+    ///
+    /// You may want to use `Foo` with [`Bar`].
+    ///
+    /// [`Bar`]: ../bar/struct.Bar.html
+    pub struct Foo;
+}
+
+pub mod bar {
+    /// Some docs for `Bar`
+    ///
+    /// You may want to use `Bar` with [`Foo`].
+    ///
+    /// [`Foo`]: ../foo/struct.Foo.html
+    pub struct Bar;
+}
+```
+
+Note that we've also had to use relative links, so that this works offline.
+Not only is this process tedious, and error prone, but it's also just wrong
+in places. If we put a `pub use bar::Bar` in our crate root, that would
+re-export `Bar` in our root. Now our links are wrong. But if we fix them,
+then they end up being wrong when we navigate to the `Bar` that lives inside
+the module. You can't actually write these links by hand, and have them all
+be accurate.
+
+In this release, you can use some syntax to let rustdoc know that you're
+trying to link to a type, and it will generate the URLs for you. Here's
+two different examples, based off of our code before:
+
+```rust
+pub mod foo {
+    /// Some docs for `Foo`
+    ///
+    /// You may want to use `Foo` with [Bar](crate::bar::Bar).
+    pub struct Foo;
+}
+
+pub mod bar {
+    /// Some docs for `Bar`
+    ///
+    /// You may want to use `Bar` with [`crate::foo::Foo`].
+    pub struct Bar;
+}
+```
+
+The first example will show the same text as before; but generate the proper
+link to the `Bar` type. The second will link to `Foo`, but will show the whole
+`crate::foo::Foo` as the link text.
+
+There are a bunch of options you can use here. Please see the ["Linking to
+items by
+name"](https://doc.rust-lang.org/stable/rustdoc/linking-to-items-by-name.html)
+section of the rustdoc book for more.
+
```

Comment:
Should we also link to the inside rust blog post about the history of this feature? https://blog.rust-lang.org/inside-rust/2020/09/17/stabilizing-intra-doc-links.html

Reply:
Sure, if folks think that's a good idea I'm not opposed.

Reply:
Added to the post!